### PR TITLE
Add support for Open Policy Agent

### PR DESCRIPTION
**Pull request by vagaerg**
## Description

This PR supersedes #17940 which was getting hard to follow due to the large number of comments

## Additional context and related issues

Docs PR by @mosabua is #20246

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(X) Release notes are required, with the following suggested text:

```markdown
## Security
* Support access control with Open Policy Agent (OPA)
```

And make it link to the docs

### CODE REVIEW

**Comment by vagaerg** [comment]
From the old PR I believe the only items that remain unresolved:

### Grants:
https://github.com/trinodb/trino/pull/17940#discussion_r1369424352

I am happy to remove this and implement a default deny if @dain or others think there is no value to supporting this in OPA.

IMHO OPA could provide a simple central point to enforce global policies as to what users can grant permissions on what catalogs, while deferring more granular permissioning to the connector access control logic for each catalog if needed.

Given that the meaning of a `GRANT` statement varies greatly across different connectors, users would need to write logic that depends on what catalog is being operated upon, but there is no reason they cannot do this with some logic in OPA.

Allowing OPA to answer an authorization question for a grant query would still let users implement their more specific permissioning logic at the connector access control level if needed, whereas denying here would immediately deny any query. It also lets users decide whether they want to globally allow or deny this style of queries.

I don't have lots of experience with this specific part of Trino, but the rest of the plugin is written with the goal of deferring all authorization requests to OPA such that users can plug in any policy they want without changing any Trino logic; and I _think_ this same goal could be beneficial here.

### Fields in the request context
https://github.com/trinodb/trino/pull/17940#discussion_r1369459481

I removed `enabledRoles` and `catalogRoles`. I have for now kept `extraCredentials` pending discussion on whether there is a cleaner way to pass in additional credentials. My original response

> We have played around (just for demos, nothing we have actually deployed) with the idea of using extra credentials to pass in JWTs that grant users extra permissions that are validated via OPA. Do you know what would be a better alternative to do this without using extraCredentials?

### Including the Trino version number in the request & documenting fields in the request

From discussion on slack with @dain 
I would like to add the Trino version to the `OpaQueryContext` object.

What way would I be able to obtain the Trino version programatically? The only way I've seen is from the `NodeManager` but that is not passed into the Access Control factory.

**This item is pending**

### Cleaning up `OpaQueryInputResource` builders

https://github.com/trinodb/trino/pull/17940#discussion_r1369405329

I would prefer to defer this for now if @dain is OK with that

## Follow-up PR

We have 3 additional contributions we would like to make, happy to open PRs after this is merged:

- Addressing the `OpaQueryInputResource` building mentioned by @dain 
- Implementing row level filtering and column masking
- Adding more options for authenticating against the OPA endpoint: OPA can be configured to require clients authenticate against it with basic credentials or mTLS ([docs](https://www.openpolicyagent.org/docs/latest/security/))

**Comment by dain** [comment]
For `extraCredentials`, my main concern passing through all extra credentials.  My suggestion is that we leave them out for now, but if we want to add them in the future, we add a configuration property for which one to include.

For version, I put up a PR to add it to `SystemAccessControlContext` #19585

**Comment by vagaerg** [comment]
> For `extraCredentials`, my main concern passing through all extra credentials. My suggestion is that we leave them out for now, but if we want to add them in the future, we add a configuration property for which one to include.

Sounds good, will remove them for now

> For version, I put up a PR to add it to `SystemAccessControlContext` #19585

That would be great! The PR looks good, thank you!

File: plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaHttpClient.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        }
+        return response.getValue();
+    }
+
+    private static String tryConvertBytesToString(byte[] bytes)
+    {
+        try {
+            return new String(bytes, UTF_8);
+        }
+        catch (Exception e) {
```
Comment (electrum) [inline plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaHttpClient.java:217]: No need to catch here. Per the specification:

> This method always replaces malformed-input and unmappable-character sequences with this charset's default replacement string.

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/HttpClientUtils.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static com.google.common.net.MediaType.JSON_UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class HttpClientUtils
+{
+    private HttpClientUtils() {}
+
+    public static class RecordingHttpProcessor
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/HttpClientUtils.java:39]: This class isn't thread safe because the setHandler method is not synchronized.  Generally the solution would be to synchronize it. And Also, I'm use `this` as the monitor instead of `this.requests`, but I think this should be change to no have a mutable handle...

Looking at the useage the three unit test classes are all constructing an instance of this class in the `@BeforeEach`, and then each test method is setting the handler.  Since this is a mock http client it is really cheap to construct and does not need to be shared.  So instead of using before/after each at all, I suggest you have the test method create the http client with the actual desired handler.  You can use helper method to create both the HTTP client and the OpaAccessControl.  If necessary you can use try-with-resouces to cleanup, but I'm not sure that is needed.  This would mean that this class could be mostly immutable (except requests), and the tests would be stateless which makes parallelization easier for us.

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/HttpClientUtils.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            implements TestingHttpClient.Processor
+    {
+        private final List<String> requests = new LinkedList<>();
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/HttpClientUtils.java:42]: If you make the change to make handler final, this can be `Collections.synchronizedList(new ArrayList<>());`... BTW array list is almost always more efficient than linked list and easier to debug (because it just has an array at the bottom)

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlFilteringUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+        }
+    }
+
+    @Test
+    public void testFilterViewQueryOwnedBy()
+    {
+        Identity userOne = Identity.ofUser("user-one");
+        Identity userTwo = Identity.ofUser("user-two");
+        List<Identity> requestedIdentities = ImmutableList.of(userOne, userTwo);
+        this.mockClient.setHandler(buildHandler("/input/action/resource/user/user", "user-one"));
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlFilteringUnitTest.java:88]: `this.` isn't needed on any of the references to `mockClient` (same in other tests)

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlFilteringUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+import static io.trino.plugin.opa.RequestTestUtilities.assertStringRequestsEqual;
+import static io.trino.plugin.opa.TestHelpers.NO_ACCESS_RESPONSE;
+import static io.trino.plugin.opa.TestHelpers.OK_RESPONSE;
+import static io.trino.plugin.opa.TestHelpers.systemSecurityContextFromIdentity;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class OpaAccessControlFilteringUnitTest
```
[the remark this answers is not in the document]
Comment (dain) [reply plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlFilteringUnitTest.java:54]: The naming convention for tests in Trino is `TestTheThing` so this should be `TestOpaAccessControlFiltering`

**Review by dain — COMMENTED** [review]
Some comments

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/RequestTestUtilities.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                    try {
+                        JsonNode parsed = jsonMapper.readTree(actualRequest);
+                        if (extractPath != null) {
+                            return parsed.at(extractPath);
+                        }
+                        return parsed;
+                    }
+                    catch (IOException e) {
+                        fail("Could not parse request", e);
+                        return null;
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/RequestTestUtilities.java:65]:
```suggestion
throw new AssertionError("Could not parse request", e);
```

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/ResponseTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+{
+    private JsonCodec<OpaQueryResult> responseCodec;
+    private JsonCodec<OpaBatchQueryResult> batchResponseCodec;
+
+    @BeforeEach
+    public void setupParser()
+    {
+        this.responseCodec = new JsonCodecFactory().jsonCodec(OpaQueryResult.class);
+        this.batchResponseCodec = new JsonCodecFactory().jsonCodec(OpaBatchQueryResult.class);
+    }
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/ResponseTest.java:41]:
```suggestion
    private final JsonCodec<OpaQueryResult> responseCodec = new JsonCodecFactory().jsonCodec(OpaQueryResult.class);
    private final JsonCodec<OpaBatchQueryResult> batchResponseCodec = new JsonCodecFactory().jsonCodec(OpaBatchQueryResult.class);
```

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/ResponseTest.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    public void testCanDeserializeOpaSingleResponse(boolean response)
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/ResponseTest.java:45]: Avoid `@ParameterizedTest`, but definitely don't use it with a boolean.  Just do this.

```suggestion
    @Test
    public void testCanDeserializeOpaSingleResponse()
    {
        testCanDeserializeOpaSingleResponse(true);
        testCanDeserializeOpaSingleResponse(false);
    }

    private void testCanDeserializeOpaSingleResponse(boolean response)
```

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/ResponseTest.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    public void testUndefinedDecisionSingleResponseTreatedAsDeny()
+    {
+        OpaQueryResult result = this.responseCodec.fromJson("{}");
+        assertFalse(result.result());
+        assertNull(result.decisionId());
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"{}", "{\"result\": []}"})
+    public void testEmptyOrUndefinedResponses(String response)
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/ResponseTest.java:90]: just inline these two cases like the boolean ones

**Review by dain — COMMENTED** [review]
Some comments

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlFilteringUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+                    }""".formatted(function.getSchemaName(), function.getFunctionName()))
+                .collect(toImmutableList());
+        assertStringRequestsEqual(expectedRequests, this.mockClient.getRequests(), "/input/action");
+    }
+
+    @ParameterizedTest(name = "{index}: {0}")
+    @MethodSource("io.trino.plugin.opa.FilteringTestHelpers#emptyInputTestCases")
+    public void testEmptyRequests(
+            BiFunction<OpaAccessControl, SystemSecurityContext, Collection> callable)
+    {
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlFilteringUnitTest.java:317]: We have been removing Parameterized tests throughtout the code base because they are difficult to understand and debug because the test cases are separated from the test code, and they are often over abstracted into generic functions like this, so a reader can not know what is actually being tested.  Please just inline the cases.

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlFilteringUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+                        }
+                    }""".formatted(function.getSchemaName(), function.getFunctionName()))
+                .collect(toImmutableList());
+        assertStringRequestsEqual(expectedRequests, this.mockClient.getRequests(), "/input/action");
+    }
+
+    @ParameterizedTest(name = "{index}: {0}")
+    @MethodSource("io.trino.plugin.opa.FilteringTestHelpers#emptyInputTestCases")
+    public void testEmptyRequests(
+            BiFunction<OpaAccessControl, SystemSecurityContext, Collection> callable)
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlFilteringUnitTest.java:316]: The `Collection` type here is a raw type which disables type checking in Java.  Add something `<?>`, or even better set the actual type.

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlSystemTest.java [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+@Testcontainers
+@TestInstance(PER_CLASS)
+public class OpaAccessControlSystemTest
+{
+    private URI opaServerUri;
+    private DistributedQueryRunner runner;
+
+    private static final int OPA_PORT = 8181;
+    @Container
+    public static GenericContainer<?> opaContainer = new GenericContainer<>(DockerImageName.parse("openpolicyagent/opa:latest-rootless"))
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlSystemTest.java:67]: The other uses of `@Container` in Trino mark the field as `private static final`, so I suggest we do that here.  In the case the field should be renamed `OPA_CONTAINER`

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlSystemTest.java [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+
+                    allow {
+                        input.context.identity.user in ["someone", "admin"]
+                    }
+                    """);
+            RuntimeException error = assertThrows(RuntimeException.class, () -> {
+                runner.execute(user("bob"), "SHOW CATALOGS");
+            });
+            assertTrue(error.getMessage().contains("Access Denied"),
+                    "Error must mention 'Access Denied': " + error.getMessage());
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlSystemTest.java:143]:
```suggestion
            assertThatThrownBy(() -> runner.execute(user("bob"), "SHOW CATALOGS"))
                    .isInstanceOf(RuntimeException.class)
                    .hasMessageContaining("Access Denied");
```

Comment (dain) [reply plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlSystemTest.java:143]: Use assertThatThrownBy in all of the tests for failure checking

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlUnitTest.java [hunk 14]
```diff
@@ -0,0 +1,1459 @@
...
+    private OpaAccessControl authorizer;
+    private final JsonMapper jsonMapper = new JsonMapper();
+    private Identity requestingIdentity;
+    private SystemSecurityContext requestingSecurityContext;
+
+    @BeforeEach
+    public void setupAuthorizer()
+    {
+        this.mockClient = new InstrumentedHttpClient(OPA_SERVER_URI, "POST", JSON_UTF_8.toString(), request -> OK_RESPONSE);
+        this.authorizer = (OpaAccessControl) new OpaAccessControlFactory().create(ImmutableMap.of("opa.policy.uri", OPA_SERVER_URI.toString()), Optional.of(mockClient));
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlUnitTest.java:77]: This create method is a static factory method, so you don't need the `new`
```suggestion
        this.authorizer = (OpaAccessControl) OpaAccessControlFactory.create(ImmutableMap.of("opa.policy.uri", OPA_SERVER_URI.toString()), Optional.of(mockClient));
```

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlUnitTest.java [hunk 15]
```diff
@@ -0,0 +1,1459 @@
...
+                    "decision_id": "foo",
+                    "some_debug_info": {"test": ""}
+                }
+                """,
+                200));
+        authorizer.checkCanShowRoles(requestingSecurityContext);
+    }
+
+    @ParameterizedTest(name = "{index}: {0}")
+    @MethodSource("io.trino.plugin.opa.OpaAccessControlUnitTest#noResourceActionTestCases")
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlUnitTest.java:107]: as with the other tests, remove use of `@ParameterizedTest`

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,1459 @@
...
+                        authorizer,
+                        requestingSecurityContext,
+                        new CatalogSchemaTableName("my_catalog", "my_schema", "my_table")));
+        assertTrue(actualError.getMessage().contains(expectedErrorMessage),
+                String.format("Error must contain '%s': %s", expectedErrorMessage, actualError.getMessage()));
+    }
+
+    private static Stream<Arguments> tableWithPropertiesTestCases()
+    {
+        Stream<FunctionalHelpers.Consumer4<OpaAccessControl, SystemSecurityContext, CatalogSchemaTableName, Map>> methods = Stream.of(
```
Comment (dain) [inline plugin/trino-opa/src/test/java/io/trino/plugin/opa/OpaAccessControlUnitTest.java:230]: Raw `Map` type

**Review by dain — COMMENTED** [review]
Looking good

File: plugin/trino-opa/src/main/java/io/trino/plugin/opa/schema/OpaBatchQueryResult.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNullElse;
+
+public record OpaBatchQueryResult(@JsonProperty("decision_id") String decisionId, @NotNull List<Integer> result)
+{
+    public OpaBatchQueryResult
+    {
+        result = ImmutableList.copyOf(requireNonNullElse(result, ImmutableList.of()));
```
[the remark this answers is not in the document]
Comment (vagaerg) [reply plugin/trino-opa/src/main/java/io/trino/plugin/opa/schema/OpaBatchQueryResult.java:28]: Yes, this depends on whether decision logging is enabled - if it is, `decision_id` will be included in the response. It is otherwise omitted ([docs](https://www.openpolicyagent.org/docs/latest/management-decision-logs/))

OPA generates its own decision_id which can be queried by the observability APIs ([docs](https://docs.styra.com/das/observability-and-audit/decision-logs/overview)). Users may also pass in one if they so wish for things like Open Tracing (we're not doing this as of this PR) https://github.com/open-policy-agent/opa/issues/1451#issuecomment-498950686

File: plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaHttpClient.java [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+                    "Sending OPA request to URI \"%s\" ; request body = %s ; request headers = %s",
+                    uri.toString(),
+                    tryConvertBytesToString(requestBodyGenerator.getBody()),
+                    request.getHeaders());
+        }
+        return FluentFuture.from(httpClient.executeAsync(request, createFullJsonResponseHandler(deserializer)))
+                .transform(response -> parseOpaResponse(response, uri), executor);
+    }
+
+    public <T> T consumeOpaResponse(ListenableFuture<T> opaResponseFuture)
```
[the remark this answers is not in the document]
Comment (vagaerg) [reply plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaHttpClient.java:107]: This is used from `OpaHighLevelClient`

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/FilteringTestHelpers.java [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+import org.junit.jupiter.api.Named;
+import org.junit.jupiter.params.provider.Arguments;
+
+import java.util.Collection;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+
+import static io.trino.plugin.opa.TestHelpers.createIllegalResponseTestCases;
+
+public class FilteringTestHelpers
```
[the remark this answers is not in the document]
Comment (vagaerg) [reply plugin/trino-opa/src/test/java/io/trino/plugin/opa/FilteringTestHelpers.java:32]: The methods in this class are used as method sources for the parametrized tests. I'll clean this up when removing the parametrized tests

File: plugin/trino-opa/src/test/java/io/trino/plugin/opa/FilteringTestHelpers.java [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+{
+    private FilteringTestHelpers() {}
+
+    public static Stream<Arguments> emptyInputTestCases()
+    {
+        Stream<BiFunction<OpaAccessControl, SystemSecurityContext, Collection>> callables = Stream.of(
```
[the remark this answers is not in the document]
Comment (vagaerg) [reply plugin/trino-opa/src/test/java/io/trino/plugin/opa/FilteringTestHelpers.java]: I will be addressing these as part of undoing the test parametrization

**Comment by vagaerg** [comment]
All comments should be addressed, I've removed parameterization on some tests, particularly the following ones in `TestOpaAccessControl`:
- `testNoResourceAction`
- `testFunctionResourceActions`
- `testCanExecuteTableProcedure`

The new `assertAccessControlMethodBehaviour` method will assert the authorizer method under test allows/denies correctly depending on the response returned by OPA and that it throws if an illegal response is returned. This method checks the request being sent to OPA,

If you are happy with the changes made, I will change all the other tests accordingly

**Comment by Cerebus** [comment]
Late to the party, but how is this performant at scale, particularly with the follow-on PR for row-based access control?  It seems to me that a REST call per row in the input is going to really stink.

Why not use the Compile API to pre-fill user and context into the policies, then execute the query with the data locally?

**Comment by vagaerg** [comment]
That is a good question. From our experimentation it seems to be working well enough as we’re deploying it alongside each worker.
The OPA binary is very simple to deploy and run so we just co locate it.

That said, I’d love to hear your suggestion on the compile API, I’m not familiar with it. Would this allow us to compile the rego and execute it externally to trino?

The row level filtering I wouldn’t expect to be much of an issue, as the actual rows are never sent to OPA. OPA just produces filter expressions that Trino applies locally - kind of like adding an extra WHERE clause

Thanks!
___________________
Pablo Arteaga

This message was sent from a mobile device, apologies for any typos or inadvertent auto-corrects.

On 14 Dec 2023, at 21:55, T. Miller ***@***.***> wrote:

﻿

Late to the party, but how is this performant at scale, particularly with the follow-on PR for row-based access control? It seems to me that a REST call per row in the input is going to really stink.

Why not use the Compile API to pre-fill user and context into the policies, then execute the query with the data locally?

**Comment by vagaerg** [comment]
> Late to the party, but how is this performant at scale, particularly with the follow-on PR for row-based access control? It seems to me that a REST call per row in the input is going to really stink.
> 
> Why not use the Compile API to pre-fill user and context into the policies, then execute the query with the data locally? E.g., https://blog.openpolicyagent.org/write-policy-in-opa-enforce-policy-in-sql-d9d24db93bf4

Just saw your updated comment with the link, thanks!
That certainly looks interesting, and perhaps something to consider in the future.

IMHO, at the moment we should keep a simple solution where OPA makes a full, final, decision.
This is mostly because Trino has no more context than what is sent to OPA: We are serializing the entirety of the available information that Trino passes down to the SPI, which isn't that large.

In fact, Trino doesn't really know all that much about the data either: in the example you linked, the application contacting OPA is aware of what pets are owned by whom and what clinicians are allowed to interact with them. This is not the case in Trino - many of these are coming from business logic that may be entirely external to Trino, and as such may as well be pushed into OPA.

E.g.: the information as to who owns a table is only partially available in Trino - tables may well be created by scripts and thus be owned by fake service accounts.

That said, _I think_ it would be nice to in the future use this compile logic to allow OPA to return expressions that Trino would execute on the data and make the decision internally. The reason I think we should defer that is two-fold:
- This is a first cut of an OPA authorizer: fully externalizing the decision making is what is most in line with the OPA design principles, and also what we _believe_ would cover a majority of users
- The complexity involved with using the compile API: turning a JSON blob into something that we would execute on top of the data internally is rather tricky, and raises all sorts of security & design questions

As for the row level filtering, note that row level filtering is done within Trino. There is never a case where each row is sent to OPA, because the OPA plugin never even sees the rows.

The Trino SPI expects authorizers to return a list of `ViewExpression` objects ([code](https://github.com/trinodb/trino/blob/3aea15e26d51f555ea8a78cc073a80bb1564f9ef/core/trino-spi/src/main/java/io/trino/spi/security/ViewExpression.java)) that contain expressions that are applied internally to further refine the results - kind of like adding an extra `WHERE` clause.

And, in fact, using row level filtering would bring us one step closer to the system you suggested 😄

**Comment by vagaerg** [comment]
I rebased the PR and bumped the version.
There's a draft PR for row level filtering & column masking here: bloomberg/trino#16 , I will bring it over once we get this one merged. That PR also adds some extra system tests.

I will also update the documentation branch (I'm around halfway through with that) and we can merge that after this PR too

**Comment by findepi** [comment]
cc @trinodb/maintainers for new plugin

File: plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaConfig.java [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+        return this;
+    }
+
+    public boolean getAllowPermissioningOperations()
+    {
+        return this.allowPermissioningOperations;
+    }
+
+    @Config("opa.allow-permissioning-operations")
+    @ConfigDescription("Whether to allow permissioning operations (GRANT, DENY, ...) as well as role management - OPA will not be queried for any such operations, they will be bulk allowed or denied depending on this setting")
```
Comment (mosabua) [inline plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaConfig.java:91]: Is this term "permissioning" really what we want to use. Is that some sort of weird terminology used in OPA .. if not I would prefer we use something like
`allow-security-operations` or `allow-sql-security-operations` or `allow-sql-security-statements` .. wdyt ?

**Comment by shohamyamin** [comment]
@vagaerg @mosabua any progress with the pr?

**Comment by soenkeliebau** [comment]
I just had to look up the link for something else as well and was wondering about the status :)

Is it maybe worth getting together for two hours one evening and hashing out the remaining open points? I'd volunteer to write minutes and post them here so everything is documented and public..
@vagaerg @mosabua @dain ?

**Review by dain — COMMENTED** [review]
This is good.  I have one comment about changing the work "permissioning", and there is a trivial merge conflict for the main pom.

Can you update these two items, squash and push, and I'll merge it?

**Comment by mosabua** [comment]
Sounds good @dain .. once this PR is merged and the details are therefore settled I will update the docs PR to the current status and ask for reviews again from @vagaerg and others.

**Comment by vagaerg** [comment]
> I just had to look up the link for something else as well and was wondering about the status :)
> 
> Is it maybe worth getting together for two hours one evening and hashing out the remaining open points? I'd volunteer to write minutes and post them here so everything is documented and public.. @vagaerg @mosabua @dain ?

Absolutely, sorry I missed this comment - I am happy to meet up whether for this PR (which will hopefully be merged soon!) or for the follow ups for row level filtering / column masking :) 

Thanks a lot for your offer @soenkeliebau !

**Comment by vagaerg** [comment]
Rebase and renaming of the field is now done @dain 
I will open a small PR shortly after this is merged to remove all parameterization. I would rather get this one merged soon and follow up on a separate PR if that's fine by you.

Thanks!

**Comment by sbernauer** [comment]
Many many thanks @vagaerg for your perseverance and @dain for merging!
